// acs/lib/command/logcat.js
var u = require('../util.js'),
    logger = require('../logger.js'),
    program = require('commander'),
    fs = require('fs'),
    path = require('path'),
    request = require('request'),
    _ = require('underscore')._,
    zlib = require('zlib'),
    util = require('util'),
    timer,
    lastLogDateTimeStamp = 'NODATETME';

exports.run = function(args, opts) {
    // console.log('Retrieves the latest log files from the server');

    var globalConfig = u.getGlobalConfig(),
        defaultConfig = u.getDefaultConfig(),
        proxy = u.getProxy();

    if(proxy) {
        logger.debug('ACS CLI is using proxy: ' + proxy.server);
        proxy = proxy.url;
    }
    if(globalConfig.publishHost && !opts.isCallback) {
        console.log(String(('Admin Hostname: ' + globalConfig.publishHost).grey));
    }

    var cookie = globalConfig.cookie ? String(globalConfig.cookie).split(';')[0] || [ 'connect-sid', '' ].join('=') : '';

	var name = null;
	if(opts.show_serverid) {
		name = opts.show_serverid === true ? null : opts.show_serverid;
	}

	var appname = u.getAppName(args, opts, name);

    /**
     * Nettle will fetch logs periodically by the interval intervalToGetLogs
     */
    var intervalToGetLogs = 5000; //the interval time is 5000 milseconds by default.

    try {
        if (opts.interval) {
            var re = /^[0-9]+.?[0-9]*$/;
            if (!re.test(opts.interval)) {
                u.die('Please enter a valid interval!'.red);
            } else if (opts.interval >= 5) {
                intervalToGetLogs = opts.interval * 1000;
            }
        }
    } catch (error) {
        u.die('Error occured:' + error);
    }

    var body = {};

    // Get server id variable
    var serverid = (opts.server_id === undefined ? 'NOSERVERID' : opts.server_id);

    var reqpath;
    if (appname) {
        reqpath = '/log/' + appname + '/' + serverid + '/';
    }

    function getLogs() {

        var url = (globalConfig.publishHost || defaultConfig.publishHost) + ':'
            + (globalConfig.publishPort || defaultConfig.publishPort) + reqpath + lastLogDateTimeStamp + '/NODATETME/LATEST/100';

        if(opts.org) {
            url += ('?orgid=' + opts.org);
        }

        var rawResBody;
        request({
            uri: url,
            method: 'POST',
            proxy: proxy,
            encoding: null,
            headers: {
                'Cookie' : cookie,
                'Content-Type': 'application/json',
                'accept-encoding' : 'gzip'
            },
            body: JSON.stringify(body)
        }, function(error, response, body){
            if (error) {
                u.die(u.getConnectionErrorMsg(error, body));
            }
            if (response.statusCode == 503) {
                u.die("Admin server is currently unavailable.");
            }
            rawResBody = body;
            zlib.gunzip(body, parseResult);
        });

        var parseResult = function(err, unzippedBody) {

            var resbody = unzippedBody;
            if(err) {
                if(err.code == 'Z_DATA_ERROR') { //in case the response body is not zipped
                    resbody = rawResBody;
                } else {
                    u.die(err);
                }
            }

            var result = {};
            try {
                result = JSON.parse(resbody);
            } catch (e) {
                logger.error(resbody);
                return;
            }
            if( result.errcode && result.errcode == 100 ) {
                clearInterval(timer);
                u.requireLogin(function() {
                    process.stdin.destroy();
                    opts.isCallback = true;
                    require('./logcat').run(args, opts);
                });
            } else {
                if (!result.success) {
                    u.die(result.message || util.inspect(result), result.errcode?result.errcode:1);
                    return;
                }
                var len = result.logmessages.length;
                for ( var i = len-1; i >= 0; i--) {
                    var msg = result.logmessages[i].message;
                    var d = msg.length - '\n'.length;
                    if(d >= 0 && msg.lastIndexOf('\n') === d)
                        msg = msg.substring(0, d);
                    console.log(((opts.show_serverid && result.logmessages[i].serverid)?'['+result.logmessages[i].serverid+'] ':'') + msg);
                }
                if (len > 0) {
                    lastLogDateTimeStamp = result.logmessages[0].time;
                }
            }
        };

    }

    getLogs();
    timer = setInterval(getLogs, intervalToGetLogs);
};
